src/tflite_cpu.cpp

```diff
@@ -0,0 +1,756 @@
+            std::unique_lock<std::mutex> interpreter_lock;
+            named_tensor_views views;
+        };
+        auto inference_result = std::make_shared<inference_result_type>();
+
+        // Walk the outputs per our metadata and emplace an
+        // appropriately typed tensor_view aliasing the interpreter
+        // output tensor buffer into the inference results.
+        for (const auto& output : state->metadata.outputs) {
+            const auto where = state->output_tensor_indices_by_name.find(output.name);
+            if (where == state->output_tensor_indices_by_name.end()) {
+                continue;  // Should be impossible
+            }
+            const auto* const tflite_tensor =
+                TfLiteInterpreterGetOutputTensor(state->interpreter.get(), where->second);
+            inference_result->views.emplace(output.name,
+                                            std::move(make_tensor_view_(output, tflite_tensor)));
+        }
+
+        // The views created in the loop above are only valid until
+        // the interpreter lock is released, so we keep the lock held
+        // by moving the unique_lock into the inference_result
+        // object. The lock itself is only valid as long as the state
+        // is alive, so we move that in too.
+        inference_result->state = std::move(state);
+        inference_result->interpreter_lock = std::move(lock);
+
+        // Finally, construct an aliasing shared_ptr which appears to
+        // the caller as a shared_ptr to views, but in fact manages
+        // the lifetime of the inference_result. When the
+        // inference_result object is destroyed, the lock will be
+        // released and the next caller can invoke the interpreter.
+        auto* const views = &inference_result->views;
+        // NOLINTNEXTLINE(performance-move-const-arg): C++20
+        return {std::move(inference_result), views};
+    }
+
+    struct metadata metadata(const vsdk::ProtoStruct& extra) final {
+        // Just return a copy of our metadata from leased state.
+        return lease_state_()->metadata;
+    }
+
+   private:
+    struct state;
+
+    void check_stopped_inlock_() const {
+        if (stopped_) {
+            std::ostringstream buffer;
+            buffer << service_name << ": service is stopped: ";
+            throw std::runtime_error(buffer.str());
+        }
+    }
+
+    std::shared_ptr<state> lease_state_() {
+        // Wait for our state to be valid or stopped and then obtain a
+        // shared_ptr to state if valid, incrementing the refcount, or
+        // throws if the service is stopped. We don't need to deal
+        // with interruption or shutdown because the gRPC layer will
+        // drain requests during shutdown, so it shouldn't be possible
+        // for callers to get stuck here.
+        std::unique_lock<std::mutex> lock(state_lock_);
+        state_ready_.wait(lock, [this]() { return (state_ != nullptr) && !stopped_; });
```

Comment:
Same issue here about hanging forever if the component is stopped.

Reply:
I'll comment on the other one.